datadog_lambda/cold_start.py

```diff
@@ -6,17 +6,28 @@
 logger = logging.getLogger(__name__)
 
 _cold_start = True
+_proactive_initialization = False
 _lambda_container_initialized = False
 
 
-def set_cold_start():
+def set_cold_start(init_timestamp_ns):
     """Set the value of the cold start global
 
     This should be executed once per Lambda execution before the execution
     """
     global _cold_start
     global _lambda_container_initialized
-    _cold_start = not _lambda_container_initialized
+    global _proactive_initialization
+    if not _lambda_container_initialized:
+        now = time.time_ns()
+        if (now - init_timestamp_ns) // 1_000_000_000 > 10:
+            _cold_start = False
```

Comment:
Hey @astuyve @joeyzhao2018 should it also be `_cold_start = not _lambda_container_initialized`? I'm getting use cold_start stats from apms and with this change I will need to take into account new `proactive_initialization= true` to measure it like before.

Based on [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtime-environment.html#:~:text=When%20you%20use,using%20provisioned%20concurrency.) I'm not sure if in this situation cold_start should be false?

Any thoughts?

Thanks

Reply:
Hi @pvicente - We're actively working with AWS on documentation for this and I can't say more right now, but you should open a support ticket with them and tell them AJ Stuyvenberg sent you ;)

A bit more about what I can say:
With this change you'll still get `cold_start:true` if the AWS Lambda function actually experienced a cold start, like so:
<img width="1281" alt="image" src="https://github.com/DataDog/datadog-lambda-python/assets/1598537/ae36ddf9-0512-4034-b67f-7d9d1354bd13">

However if the function is proactively initialized, then it's not a cold start in the sense that there was no additional latency caused by sandbox initialization (as init occurred seconds, hours, or minutes before the first request was served by the sandbox).

We now tag those correctly, so when we use APM to select this python invocation:
<img width="1283" alt="image" src="https://github.com/DataDog/datadog-lambda-python/assets/1598537/573099c8-3650-44c6-9af1-709609eff065">
We can see that `initialization` occurred 3m59s before the function was executed:
<img width="823" alt="image" src="https://github.com/DataDog/datadog-lambda-python/assets/1598537/4445c56e-7742-46c0-9078-fa00427ce247">


I hope I've helped!

Reply:
Thanks for the explanation @astuyve 🥇